src/control/lib/telemetry/promexp/collector.go

```diff
@@ -94,28 +94,38 @@ func NewCollector(log logging.Logger, opts *CollectorOpts, sources ...*EngineSou
 	return c, nil
 }
 
-func fixPath(in string) (labels labelMap, name string) {
-	name = strings.Map(func(r rune) rune {
-		if !unicode.IsPrint(r) || unicode.IsSpace(r) {
-			return '_'
-		}
-		switch r {
-		case '/', ':':
+func sanitizeMetricName(in string) string {
+	return strings.Map(func(r rune) rune {
+		switch {
+		// Valid names for Prometheus are limited to:
+		case r >= 'a' && r <= 'z': // lowercase letters
+		case r >= 'A' && r <= 'Z': // uppercase letters
+		case unicode.IsDigit(r): // digits
+		case r == ':':
+		default: // sanitize any other character
 			return '_'
 		}
 
 		return r
 	}, strings.TrimLeft(in, "/"))
+}
+
+func fixPath(in string) (labels labelMap, name string) {
+	name = sanitizeMetricName(in)
 
 	labels = make(labelMap)
-	ID_re := regexp.MustCompile(`ID_(\d+)_`)
-	io_re := regexp.MustCompile(`io_(\d+)_`)
+	ID_re := regexp.MustCompile(`ID\:?_+(\d+)_?`)
+	io_re := regexp.MustCompile(`io_+(\d+)_?`)
 
 	name = ID_re.ReplaceAllString(name, "")
 	io_matches := io_re.FindStringSubmatch(name)
 	if len(io_matches) > 0 {
 		labels["target"] = io_matches[1]
-		name = io_re.ReplaceAllString(name, "io_")
+		replacement := "io"
```

Comment:
While you're in here fixing this stuff, could you add some rules to handle these too?
```engine_net_1_0_failed_addr           Counter Total number of failed address resolution attempts
engine_net_1_0_req_timeout           Counter Total number of timed out RPC requests
engine_net_1_0_uri_lookup_timeout    Counter Total number of timed out URI lookup requests
engine_net_1_1_failed_addr           Counter Total number of failed address resolution attempts
engine_net_1_1_req_timeout           Counter Total number of timed out RPC requests
engine_net_1_1_uri_lookup_timeout    Counter Total number of timed out URI lookup requests
engine_net_1_2_failed_addr           Counter Total number of failed address resolution attempts
engine_net_1_2_req_timeout           Counter Total number of timed out RPC requests
engine_net_1_2_uri_lookup_timeout    Counter Total number of timed out URI lookup requests
```

I believe the first number is the rank and the second is the context (@johannlombardi is this correct?). So these could be transformed into:
engine_net_failed_addr, engine_net_req_timeout, etc. with the rank/context being added as labels.

Reply:
Will fix. Wasn't sure how to parse those just off the cuff.